### PR TITLE
feat(new integration): onboarding sendinblue cloud mode destination

### PR DESCRIPTION
**Pull request by Gauravudia**
## Description of the change

> onboarding sendinblue cloud mode destination

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: src/v0/destinations/sendinblue/util.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      {
+        scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.TRANSFORMATION.SCOPE,
+        meta: TRANSFORMER_METRIC.MEASUREMENT_TYPE.TRANSFORMATION.META.BAD_PARAM
+      },
+      DESTINATION
+    );
+  }
+};
+
+const prepareEmailFromPhone = phone => {
```
Comment (shrouti1507) [inline src/v0/destinations/sendinblue/util.js:55]: can we add some context about this? Do we have any public doc regarding this? If yes, lets link that. Or if it is totally experiment basis, add that detail in comments.

File: src/v0/destinations/sendinblue/util.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
+const { DESTINATION, EMAIL_SUFFIX } = require("./config");
+const { TransformationError } = require("../../util");
+const { TRANSFORMER_METRIC } = require("../../util/constant");
+
+const validateEmail = email => {
```
Comment (shrouti1507) [inline src/v0/destinations/sendinblue/util.js:5]: move this is common utils.

File: src/v0/destinations/sendinblue/util.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  const regex = /^(([^\s"(),.:;<>@[\\\]]+(\.[^\s"(),.:;<>@[\\\]]+)*)|(".+"))@((\[(?:\d{1,3}\.){3}\d{1,3}])|(([\dA-Za-z-]+\.)+[A-Za-z]{2,}))$/;
+  return !!regex.test(email);
+};
+
+const validatePhoneWithCountryCode = phone => {
```
Comment (shrouti1507) [inline src/v0/destinations/sendinblue/util.js:10]: move this in common utils

File: src/v0/destinations/sendinblue/data/SendinblueTrackLinkConfig.json [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    "destKey": "email",
+    "sourceKeys": ["context.traits.email", "properties.email"]
+  },
+  {
+    "destKey": "phone",
+    "sourceKeys": ["context.traits.phone", "properties.phone"]
+  },
+  {
+    "destKey": "link",
+    "sourceKeys": "properties.link",
```
Comment (shrouti1507) [inline src/v0/destinations/sendinblue/data/SendinblueTrackLinkConfig.json:12]: what does this link signify. I remember we had some hiccups while implementing this in device mode. How this one is different? Can't we map context.page.url here?

Comment (Gauravudia) [reply src/v0/destinations/sendinblue/data/SendinblueTrackLinkConfig.json:12]: - Track link is not implemented in device mode.
- As, per their doc -> `Track link is used to track click events on various links of your website.` . 
- I think we can't map `context.page.url` to `link`.  As this is page url.

File: src/v0/destinations/sendinblue/util.js (hunk 1 above)
Comment (Gauravudia) [reply src/v0/destinations/sendinblue/util.js:55]: This is experimental. Sendinblue is appending `@mailin-sms.com` to `phone` number and setting it in a `email` field if only phone is passed while creating a contact.

File: src/v0/destinations/sendinblue/data/SendinblueTrackEventConfig.json [hunk 5]
```diff
@@ -0,0 +1,30 @@
+[
+  {
+    "destKey": "email",
+    "sourceKeys": ["context.traits.email", "properties.email"]
+  },
+  {
+    "destKey": "phone",
+    "sourceKeys": ["context.traits.phone", "properties.phone"]
```
[the remark this answers is not in the document]
Comment (Gauravudia) [reply src/v0/destinations/sendinblue/data/SendinblueTrackEventConfig.json]: updated

File: src/v0/destinations/sendinblue/data/SendinblueTrackEventConfig.json [hunk 6]
```diff
@@ -0,0 +1,30 @@
+[
+  {
+    "destKey": "email",
+    "sourceKeys": ["context.traits.email", "properties.email"]
```
[the remark this answers is not in the document]
Comment (Gauravudia) [reply src/v0/destinations/sendinblue/data/SendinblueTrackEventConfig.json]: updated

File: src/v0/destinations/sendinblue/data/SendinblueTrackLinkConfig.json [hunk 7]
```diff
@@ -0,0 +1,22 @@
+[
+  {
+    "destKey": "email",
+    "sourceKeys": ["context.traits.email", "properties.email"]
```
[the remark this answers is not in the document]
Comment (Gauravudia) [reply src/v0/destinations/sendinblue/data/SendinblueTrackLinkConfig.json]: updated

File: src/v0/destinations/sendinblue/data/SendinblueTrackLinkConfig.json [hunk 8]
```diff
@@ -0,0 +1,22 @@
...
+  },
+  {
+    "destKey": "phone",
```
[the remark this answers is not in the document]
Comment (Gauravudia) [reply src/v0/destinations/sendinblue/data/SendinblueTrackLinkConfig.json]: updated

File: src/v0/destinations/sendinblue/transform.js [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+// ref:- https://developers.sendinblue.com/reference/createcontact
+const createOrUpdateContactResponseBuilder = (message, destination) => {
+  const { endpoint } = CONFIG_CATEGORIES.CREATE_OR_UPDATE_CONTACT;
+  const payload = constructPayload(
+    message,
+    MAPPING_CONFIG[CONFIG_CATEGORIES.CREATE_OR_UPDATE_CONTACT.name]
+  );
+
+  checkIfEmailOrPhoneExists(payload.email, payload.attributes?.SMS);
+  validateEmailAndPhone(payload.email, payload.attributes?.SMS);
```
Comment (ItsSudip) [inline src/v0/destinations/sendinblue/transform.js:67]: Can't we merge these two functions?

File: src/v0/destinations/sendinblue/data/SendinbluePageConfig.json [hunk 10]
```diff
@@ -0,0 +1,39 @@
+[
+  {
+    "destKey": "email",
+    "sourceKeys": "emailOnly",
+    "sourceFromGenericMap": true
+  },
+  {
+    "destKey": "ma_title",
```
Comment (ItsSudip) [inline src/v0/destinations/sendinblue/data/SendinbluePageConfig.json:8]:
```suggestion
    "destKey": "properties.ma_title",
```
We can also provide as above.

**Review by ItsSudip — CHANGES_REQUESTED** [review]

File: src/v0/destinations/sendinblue/transform.js (hunk 9 above)
Comment (Gauravudia) [reply src/v0/destinations/sendinblue/transform.js:113]: In some places `email` is mandatory where we don't need to check if `phone` is exists in input or not. For that we can skip the use of 
`checkIfEmailOrPhoneExists` and can simply validate email/phone with `validateEmailAndPhone`. Thats why i have created these two separate functions.

File: src/v0/destinations/sendinblue/util.js (hunk 2 above)
Comment (Gauravudia) [reply src/v0/destinations/sendinblue/util.js]: moved in common utils

File: src/v0/destinations/sendinblue/util.js (hunk 3 above)
Comment (Gauravudia) [reply src/v0/destinations/sendinblue/util.js]: moved in common utils

File: src/v0/destinations/sendinblue/transform.js (hunk 9 above)
Comment (ItsSudip) [reply src/v0/destinations/sendinblue/transform.js:113]: We can create one function and provide some flags to differentiate between these two functionalities.

File: src/v0/util/index.js [hunk 11]
```diff
@@ -1109,6 +1109,28 @@ function getDestinationExternalID(message, type) {
   return destinationExternalId;
 }
 
+/**
+ * Function to get the the externalId array with the given type
+ * @param {*} message {"context":{"externalId":[{"type":"customerListId","id":1},{"type":"customerListId","id":2}]}}
```
Comment (ItsSudip) [inline src/v0/util/index.js:1114]: It is not correct. We should take an array as id.
{"context":{"externalId":[{"type":"customerListId","id":[1,2]}]}}

**Review by ItsSudip — CHANGES_REQUESTED** [review]

File: test/__mocks__/data/sendinblue/response.json [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+    "modifiedAt": "2022-12-18T14:06:20.515+05:30",
+    "attributes": {
+      "LASTNAME": "Pittman",
+      "FIRSTNAME": "Gordon",
+      "AGE": 33
+    },
+    "listIds": [5],
+    "statistics": {}
+  }
+}
```
[the remark this answers is not in the document]
Comment (shrouti1507) [reply test/__mocks__/data/sendinblue/response.json:32]: +1

File: src/v0/destinations/sendinblue/util.js (hunk 1 above)
Comment (shrouti1507) [reply src/v0/destinations/sendinblue/util.js:80]: then please add that in comments. Otherwise it will look confusing.

Comment (Gauravudia) [reply src/v0/destinations/sendinblue/util.js:80]: comment added

File: src/v0/destinations/sendinblue/transform.js (hunk 9 above)
Comment (Gauravudia) [reply src/v0/destinations/sendinblue/transform.js:113]: Yes, 2 separate function with different functionality

File: src/v0/destinations/sendinblue/data/SendinblueCreateOrUpdateContactConfig.json [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+  },
+  {
+    "destKey": "attributes",
+    "sourceKeys": "traits",
+    "sourceFromGenericMap": true
+  },
+  {
+    "destKey": "attributes.EMAIL",
+    "sourceKeys": ["traits.newEmail", "context.traits.newEmail"]
+  },
```
[the remark this answers is not in the document]
Comment (Gauravudia) [reply src/v0/destinations/sendinblue/data/SendinblueCreateOrUpdateContactConfig.json:15]: - In CreateDOIContactConfig `email` is mandatory. Whereas in CreateOrUpdate `email` field is non mandatory.
- `EMAIL` field inside the attribute signifies we are using it to update the existing email.
- In CreateOrUpdate we have only one api so we are updating the parameters in the same request.
- For DOI we have 2 APIs -> Create DOI and Update DOI .We are sending EMAIL inside the attributes in update DOI contact API
- I have used different configs for `normal` and `doi` contact for simplicity

File: src/v0/util/index.js (hunk 11 above)
Comment (Gauravudia) [reply src/v0/util/index.js:1114]: @shrouti1507 suggested this  in Clickup. So, i am using the same here. Let me know if i need to use array here

File: test/__mocks__/data/sendinblue/response.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (Gauravudia) [reply test/__mocks__/data/sendinblue/response.json:32]: failed case for contact not found (404) is there in `sendinblue.mock.js` file

File: src/v0/destinations/sendinblue/data/SendinbluePageConfig.json (hunk 10 above)
Comment (Gauravudia) [reply src/v0/destinations/sendinblue/data/SendinbluePageConfig.json:8]: Do i need to add this mapping for `path` , `referrer` as well?
i.e. 
`properties.ma_referrer, properties.ma_path`

File: src/v0/destinations/sendinblue/transform.js [hunk 14]
```diff
@@ -0,0 +1,459 @@
...
+        `At least one of email or phone or contactId is required to unlink the contact from a given list`,
+        400,
+        {
+          scope: TRANSFORMER_METRIC.MEASUREMENT_TYPE.TRANSFORMATION.SCOPE,
+          meta:
+            TRANSFORMER_METRIC.MEASUREMENT_TYPE.TRANSFORMATION.META.BAD_EVENT
+        },
+        DESTINATION
+      );
+    }
```
[the remark this answers is not in the document]
Comment (Gauravudia) [reply src/v0/destinations/sendinblue/transform.js]: updated

File: src/v0/util/index.js [hunk 15]
```diff
@@ -1086,7 +1086,7 @@ const constructPayload = (message, mappingJson, destinationName = null) => {
 //   "context": {
 //     "externalId": [
 //       {
-//         "type": "kustomerId",
+//         "type": "customerId",
```
[the remark this answers is not in the document]
Comment (Gauravudia) [reply src/v0/util/index.js]: changes reverted

File: src/v0/destinations/sendinblue/data/SendinbluePageConfig.json (hunk 10 above)
Comment (ItsSudip) [reply src/v0/destinations/sendinblue/data/SendinbluePageConfig.json:8]: Yes, I want to say we support multilevel mapping as well so wherever possible we can use that.

Comment (Gauravudia) [reply src/v0/destinations/sendinblue/data/SendinbluePageConfig.json:8]: added the above mapping for ma_title, ma_path, ma_referrer

File: src/v0/util/index.js (hunk 11 above)
Comment (Gauravudia) [reply src/v0/util/index.js]: updated

**Review by shrouti1507 — APPROVED** [review]

**Review by mihir-4116 — APPROVED** [review]

**Review by anantjain45823 — APPROVED** [review]

**Review by shrouti1507 — APPROVED** [review]

**Review by utsabc — APPROVED** [review]

**Pull request merged**